### ISSUE TITLE
Installijng without the need to have visual studio on windows

### ISSUE THREAD

**Issue by ParisNeo** [turboderp/exllamav2#117]
Hi.

I get many claims about the exllama2 binding, as it requires the users to have visual studio build tools preinstalled on the machine. Is there a way to avoid this? I see that in the code you require to have cl.exe if on windows.

Best regards

**Comment by ParisNeo** [comment]
Perfect. Now it is working.
Thanks for the tip!

Is there a metal version for macos users? or rocm version or vulkan for amd users?

**Comment by turboderp** [comment]
There are prebuilt ROCm 5.6 wheels, yes, but no Metal support planned yet.

**Comment by ParisNeo** [comment]
Do you have a discord channel for your project or do you have a discord acount. If you do you can come hangout on lollms's discord channel: https://discord.gg/mzfqBtEd

I'm trying to make prebuilt versions of lollms packed with prebuilt bingins. The users of lollms are mostly newbies and do not have enough technical skills to install things. I need to have a click to install script. I already have built one for each platform, but once the tool is installed, the users have to install the bindings. The bindings scripts do have install function that does the install automatically. On the UI they just press install and then wait.

I added a platform detection system to select which prebuilt files should be used and it worked on my PC but on some other users they have dll problems on windows.

Also, the models on exllama2 are very talkative. I'm sure I'm doing something wrong in the initialization or generation process as the hugging face binding on the same model works better but slower. So I think it is a parametrization problem.

Do you have an example that you know works, so I can get inspired from to update my binding code:
https://github.com/ParisNeo/lollms_bindings_zoo/blob/main/exllama2/__init__.py

And many thanks for all your answers. I think people will like the exllama2 binding once it is ready.

**Comment by turboderp** [comment]
Okay, so here are some comments I guess. Keep in mind I'm just skimming the code and might be misunderstanding a few things.

1. In `build_model` you seem to be scanning for files matching `['.safetensors', '.pt', '.bin']` but ExLlamaV2 only supports Safetensors files. It also supports sharded models (with multiple .safetensors files in the same directory).

2. As for the models being too talkative, I would still get rid of this line:

```python
self.settings.disallow_tokens(self.tokenizer, [self.tokenizer.eos_token_id])
```

Since it explicitly asks the generator to assign an infinite negative bias to the EOS token.

3. The default repetition penalty also seems rather high to me, which could steer the model away from whatever end-of-reply signal it's trying to emit (whether that's registered by the generator itself or elsewhere in your stack.) Overall though, it can really hard to pin down why you're getting "good" behavior in one instance and not the other, since there are so many variables to consider. So you want the framework to be as robust as possible and to try to clearly define exactly what constitutes a reply, why the model is replying in the first place and how it would know that it's time to stop talking.

4. The Hugging Face sampler works a little differently but isn't necessarily more correct. The differences you're seeing could be down to things like the order in which samplers are applied. E.g. top-K and top-P both take the input logits, filter out all but some subset of them, and then normalizes the probabilities afterwards. That means if top-P is evaluated after top-K, the two are not independent. If top-P is evaluated first, though, then they are. Which is the correct behavior? Which is *expected*? Idk.

5. Encoding matters. Pay attention to how strings are tokenized in your HF binding vs the ExLlama binding. For instance, the presence of a BOS token at the start of the prompt might be hinting at the model that it should end its reply with an EOS token.

6. Model can also be very sensitive to whitespace around the end of a prompt. I added token healing to the ExLlama generator which can help address that. You can try enabling it with `token_healing = True` when calling `begin_stream`.

There are some examples in this repo to look at, but I would pay attention to the chatbot especially. It works well with a number of common prompt formats, and it also has the "raw" mode which works well with base models (although base models are inherently unpredictable so YMMV).

**Comment by ParisNeo** [comment]
Thank you very much for your answer. I'll try to investigate and apply your advices.
I appreciate your help.

**Comment by ParisNeo** [comment]
I think I've done it.
I just explicitely added a BOS to the tokens I send the model and it does seem to become way less talkative this way.
Thank you very much
